backend/de.metas.async/src/main/java/de/metas/async/processor/impl/planner/QueueProcessorPlanner.java

```diff
@@ -0,0 +1,360 @@
+				}
+			}
+		}
+	}
+
+	@NonNull
+	public List<IQueueProcessor> getRegisteredQueueProcessors()
+	{
+		return ImmutableList.copyOf(queueProcessors.values());
+	}
+
+	public void addQueueProcessor(@NonNull final IQueueProcessor queueProcessor)
+	{
+		validateQueueProcessorAssignment(queueProcessor);
+
+		queueProcessors.put(queueProcessor.getQueueProcessorId(), queueProcessor);
+	}
+
+	public boolean removeQueueProcessor(@NonNull final QueueProcessorId queueProcessorId)
+	{
+		final Optional<IQueueProcessor> processorOpt = getQueueProcessor(queueProcessorId);
+
+		if (!processorOpt.isPresent())
+		{
+			return true;
+		}
+
+		processorOpt.get().shutdownExecutor();
+
+		queueProcessors.remove(processorOpt.get().getQueueProcessorId());
+
+		return true;
+	}
+
+	@NonNull
+	public Optional<IQueueProcessor> getQueueProcessor(@NonNull final QueueProcessorId queueProcessorId)
+	{
+		return Optional.ofNullable(queueProcessors.get(queueProcessorId));
+	}
+
+	public void start()
+	{
+		mainLock.lock();
+		try
+		{
+			if (isRunning.get())
+			{
+				logger.debug("QueueProcessorPlanner.start() - already running! return...");
+				return;
+			}
+
+			this.startPlanner();
+		}
+		finally
+		{
+			mainLock.unlock();
+		}
+	}
+
+	public void shutdown()
+	{
+		mainLock.lock();
+		try
+		{
+			if (!queueProcessors.isEmpty())
+			{
+				queueProcessors
+						.values()
+						.stream()
+						.map(IQueueProcessor::getQueueProcessorId)
+						.forEach(this::removeQueueProcessor);
+			}
+
+			this.shutdownPlanner();
+		}
+		finally
+		{
+			mainLock.unlock();
+		}
+	}
+
+	private boolean run0()
+	{
+		final List<IQueueProcessor> availableProcessors = queueProcessors
+				.values()
+				.stream()
+				.filter(IQueueProcessor::isAvailableToWork)
+				.collect(ImmutableList.toImmutableList());
+
+		if (availableProcessors.isEmpty())
+		{
+			logger.debug("run0 - all processors are busy! Skipping...");
+			return false;
+		}
+
+		final Properties workPackageCtx = Env.newTemporaryCtx();
+
+		final List<I_C_Queue_WorkPackage> workPackages = pollAndLockWorkPackages(workPackageCtx, availableProcessors);
+		if (workPackages.isEmpty())
+		{
+			logger.debug("pollAndLockWorkpackages - returned no workPackage to be processed for availableProcessorIds={}", availableProcessors);
+			return false;
+		}
+
+		return handleWorkPackages(workPackageCtx, workPackages);
+	}
+
+	private boolean handleWorkPackages(@NonNull final Properties ctx, @NonNull final List<I_C_Queue_WorkPackage> workPackages)
+	{
+		if (workPackages.isEmpty())
+		{
+			return false;
+		}
+
+		boolean handledAllWorkPackages = true;
+
+		for (final I_C_Queue_WorkPackage workPackage : workPackages)
+		{
+			if (!isValid(workPackage))
+			{
+				WorkPackageLockHelper.unlockNoFail(workPackage);
+				handledAllWorkPackages = false;
+				continue;
+			}
+
+			final Optional<IQueueProcessor> queueProcessorOptional = getAvailableQueueProcessorForWorkPackage(workPackage);
```

Comment:
instead of
```java
final Optional<IQueueProcessor> queueProcessorOptional = getAvailableQueueProcessorForWorkPackage(workPackage)
```

i would prefer 
```java
final IQueueProcessor queueProcessor= getAvailableQueueProcessorForWorkPackage(workPackage).orElse(null);
```

IMHO it makes the code way cleaner and light, 
but that's a personal taste.